Make atmos_fcst job work using GFSv17 and new COM structure

XianwuXue-NOAA opened the pull request:
The job works on both WCOSS2 and HARA.

XianwuXue-NOAA committed:
Move gefs_fcst.part to parm_gefs

 On branch feature/v13_atmos_fcst
	renamed:    parm/gefs_fcst.parm -> parm/parm_gefs/gefs_fcst.parm

Refs: #94

XianwuXue-NOAA committed:
Make changes based on previous changes

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   rocoto/bin/wcoss2/common.sh
	modified:   rocoto/bin/wcoss2/forecast_hr.sh
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Use preamble.sh and add lines from previous changes
 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Merge branch 'feature/v13_atm_only' into feature/v13_atmos_fcst

XianwuXue-NOAA committed:
Modify j-job of fcst to use new COM

Also modify parm files

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   parm/.gitignore
	new file:   parm/config/config.ufs
	modified:   parm/parm_gefs/gefs.parm
	modified:   parm/parm_gefs/gefs_fcst.parm
	new file:   parm/parm_gefs/gefs_fcst_resources.parm
	modified:   sorc/link_gefs.sh

Refs: #94

XianwuXue-NOAA committed:
Optimize and clean ex-script of fcst

 On branch feature/v13_atmos_fcst
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Fix some bugs and improve to run fcst job

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   parm/.gitignore
	modified:   parm/config/config.ufs
	modified:   parm/parm_gefs/gefs.parm
	modified:   parm/parm_gefs/gefs_fcst.parm
	modified:   rocoto/parm/setbase
	modified:   rocoto/py/GEFS_XML_For_Tasks.py
	modified:   rocoto/user_full.conf
	modified:   scripts/exgefs_forecast.sh
	modified:   sorc/link_gefs.sh
	modified:   ush/.gitignore

Refs: #94

XianwuXue-NOAA committed:
Remove imp_physics in the user_full.conf

 On branch feature/v13_atmos_fcst
	modified:   rocoto/user_full.conf

Refs: #94

XianwuXue-NOAA committed:
Improve and cleanup j-job, ex-script of fcst
 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   parm/parm_gefs/gefs_fcst.parm
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Improve bin file of forecast

 On branch feature/v13_atmos_fcst
	modified:   rocoto/bin/hera/forecast_hr.sh
	modified:   rocoto/bin/wcoss2/forecast_hr.sh

Refs: #94

XianwuXue-NOAA committed:
Modify to run fcst on Hera

 On branch feature/v13_atmos_fcst
	modified:   rocoto/bin/hera/forecast_hr.sh
	modified:   rocoto/parm/setbase

Refs: #94

XianwuXue-NOAA commented on parm/config/config.ufs:
```diff
@@ -0,0 +1,409 @@
+#export jchunk2d=$((2*RESTILE))
+#export ichunk3d=$((4*RESTILE))
+#export jchunk3d=$((2*RESTILE))
+#export kchunk3d=1
```
Lines 210~252 are for GEFS fcst job, especially for d16_35 and chem

XianwuXue-NOAA commented on parm/parm_gefs/gefs_fcst.parm:
```diff
@@ -0,0 +1,572 @@
+elif [[ "$imp_physics" -eq 11 ]]; then # GFDL
+    export ncld=5
+    if [[ $cplchm = ".true." ]]; then
+        export FIELD_TABLE="$HOMEgfs/parm/parm_fv3diag/chm_field_table_gfdl"
```
This is for chem, maybe we don't use it in gefs.v13

XianwuXue-NOAA commented on parm/parm_gefs/gefs_fcst.parm:
```diff
@@ -0,0 +1,572 @@
+# Disable the use of coupler.res; get model start time from model_configure
+export USE_COUPLER_RES="NO"
+
+    # Write more variables to output
```
In original config.fcst, there is no condition for `gefs`

XianwuXue-NOAA replied:
@WalterKolczynski-NOAA @aerorahul

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @aerorahul Could you please review my changes for fcst job? I have a few comments for the config files, could you also give me some suggestions? I have ran c00, p01 and p02 (C48) succeeded on Hera, but failed for p02 on WCOSS2 because of instability of the model.

XianwuXue-NOAA committed:
Modify stochastic parameters

 On branch feature/v13_atmos_fcst
	modified:   parm/parm_gefs/gefs_fcst.parm
	modified:   rocoto/user_full.conf

Refs: #94

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
-    last_cyc=$(echo $last_date | cut -c9-10)
-    export WRDIR=${WRDIR:-$(compath.py $envir/com/$NET/${ver})}/${RUN}.${last_PDY}/$last_cyc/wave/restart
+  # Set location of wave restart from last cycle
+  last_date=$($NDATE -$WAVHCYC $CDATE)
```
```suggestion
  last_date=$(${NDATE} -${WAVHCYC} "${PDY}${cyc}")
```

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
-    export WRDIR=${WRDIR:-$(compath.py $envir/com/$NET/${ver})}/${RUN}.${last_PDY}/$last_cyc/wave/restart
+  # Set location of wave restart from last cycle
+  last_date=$($NDATE -$WAVHCYC $CDATE)
+  last_PDY=$(echo $last_date | cut -c1-8)
+  last_cyc=$(echo $last_date | cut -c9-10)
```
```suggestion
  last_PDY=${last_date:0:8}
  last_cyc=${last_date:8:2}
```

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
 export COMOUT=${COMOUT:-$(compath.py -o $NET/${ver})/${RUN}.${PDY}/$cyc}
-export GESIN=${GESIN:-$(compath.py $envir/com/$NET/${ver})/${RUN}.${PDY}/$cyc/nwges}
-export GESOUT=${GESOUT:-$(compath.py -o $NET/${ver})/${RUN}.${PDY}/$cyc/nwges}
+export ROTDIR=${ROTDIR:-$(compath.py $envir/com/$NET/${ver})}
```
These should no longer be necessary with the templates. `$ROTDIR` should be defined in the job card.
```suggestion
```

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
+declare -x gcyc="${GDATE:8:2}"
+
+# Construct COM variables from templates (see config.com)
+export memchar=$(echo $RUNMEM|cut -c3-5)
```
```suggestion
declare -rx memchar=${RUNMEM:2:3}
```

WalterKolczynski-NOAA replied:
Really, we should just be moving on from p## and use mem###.

XianwuXue-NOAA replied:
OK. I will open another issue to do it. for c00->000, p01->001, how about **aer**?

WalterKolczynski-NOAA replied:
Leave it as aer for now.

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
+MEMDIR="${memchar}" YMD=${PDY} HH=${cyc} generate_com -rx COM_ATMOS_RESTART COM_ATMOS_INPUT COM_ATMOS_ANALYSIS \
+  COM_ATMOS_HISTORY COM_ATMOS_MASTER
+
+echo "COM_ATMOS_RESTART: " ${COM_ATMOS_RESTART}
+echo "COM_ATMOS_INPUT: " ${COM_ATMOS_INPUT}
+echo "COM_ATMOS_ANALYSIS: " ${COM_ATMOS_ANALYSIS}
+echo "COM_ATMOS_HISTORY: " ${COM_ATMOS_HISTORY}
+echo "COM_ATMOS_MASTER: " ${COM_ATMOS_MASTER}
```
This is now unnecessary. `generate_com()` echoes the assignements.

```suggestion
```

XianwuXue-NOAA replied:
this is just echo them for me. I will delete them now.

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
+
+RUN=${rCDUMP} MEMDIR="${memchar}" YMD="${gPDY}" HH="${gcyc}" generate_com -rx COM_ATMOS_RESTART_PREV:COM_ATMOS_RESTART_TMPL
+
+echo "COM_ATMOS_RESTART_PREV: " ${COM_ATMOS_RESTART_PREV}
```
```suggestion
```

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
+  last_date=$($NDATE -$WAVHCYC $CDATE)
+  last_PDY=$(echo $last_date | cut -c1-8)
+  last_cyc=$(echo $last_date | cut -c9-10)
+  export WRDIR=${WRDIR:-$(compath.py $envir/com/$NET/${ver})}/${RUN}.${last_PDY}/$last_cyc/wave/restart
```
```suggestion
```

XianwuXue-NOAA replied:
If you want to remove WRDIR, then I think I can remove the whole block. However, this PR is only for atmos. So I will not test these changes. I plan when I do the coupled, I can modify the other statements not related to atmos.

WalterKolczynski-NOAA replied:
You already have it farther down as `COM_WAVE_RESTART_PREV`, which is what the scripts in global-workflow will look for now (though this is untested and may need additional work).

XianwuXue-NOAA replied:
I will delete this block since it seems that we don't need it

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,96 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
 else
-    echo "FATAL ERROR in ${.sh.file}: $job failed!"
-    exit $err
+  echo "FATAL ERROR in ${BASH_SOURCE}: $job failed!"
```
```suggestion
  echo "FATAL ERROR in ${BASH_SOURCE[0]}: $job failed!"
```

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -125,6 +164,6 @@ fi
 if [ ${KEEPDATA:-NO} = NO ] ; then rm -rf $DATA ; fi
 
-echo "$(date -u) end ${.sh.file}"
+echo "$(date -u) end ${BASH_SOURCE}"
```
Not necessary with the postamble
```suggestion
```

WalterKolczynski-NOAA commented on scripts/exgefs_forecast.sh:
```diff
@@ -231,109 +188,100 @@ echo $FCSTEXEC
 if [[ $err != 0 ]]; then
-	echo "FATAL ERROR in ${.sh.file}: received a non-zero return code from $FORECASTSH"
-	exit $err
+  echo "FATAL ERROR in ${BASH_SOURCE}: received a non-zero return code from $FORECASTSH"
```
```suggestion
  echo "FATAL ERROR in ${BASH_SOURCE[0]}: received a non-zero return code from $FORECASTSH"
```

WalterKolczynski-NOAA commented on scripts/exgefs_forecast.sh:
```diff
@@ -231,109 +188,100 @@ echo $FCSTEXEC
 fi
 
-echo "$(date -u) end ${.sh.file}"
+echo "$(date -u) end ${BASH_SOURCE}"
```
```suggestion
echo "$(date -u) end ${BASH_SOURCE[0]}"
```

WalterKolczynski-NOAA commented on scripts/exgefs_forecast.sh:
Much of this should be in a config file.

XianwuXue-NOAA replied:
in config.fcst in g-w? If it is, you and @aerorahul should determine can I use `cdump=gefs`? or only use `RUN=gefs`? If you want to use run=gefs, then we should modify the config* in g-w to use RUN instead of `cdump`

WalterKolczynski-NOAA replied:
Per the slack and in-person conversations, make GEFS versions of the config files and then make sure these settings are defined in there.

WalterKolczynski-NOAA reviewed (changes requested):
Instead of duplicating the contents of GFS config files to make a few changes, you should be making changes to the config files in global workflow to support GEFS and/or adding config files that just change the things that are different. We want GEFS to run more more like GFS.

XianwuXue-NOAA committed:
Modify based on Walter's Comments

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA I have submitted a few commits based on your comment. When you have time, please review it.

XianwuXue-NOAA committed:
Rename gefs_fcst_resources.parm to config.resources

 On branch feature/v13_atmos_fcst
	renamed:    parm/parm_gefs/gefs_fcst_resources.parm -> parm/config/config.resources
	modified:   parm/parm_gefs/gefs_fcst.parm

Refs: #94

XianwuXue-NOAA committed:
Clean up some unused statements

 On branch feature/v13_atmos_fcst
	modified:   parm/parm_gefs/gefs_fcst.parm
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Rename the gefs parm files to config files

 On branch feature/v13_atmos_fcst
 	modified:   jobs/JGEFS_ATMOS_PREP
	modified:   jobs/JGEFS_FORECAST
	renamed:    parm/parm_gefs/gefs.parm -> parm/config/config.base
	renamed:    parm/parm_gefs/gefs_fcst.parm -> parm/config/config.fcst
	renamed:    parm/parm_gefs/gefs_atmos_prep.parm -> parm/config/config.prep

Refs: #94

XianwuXue-NOAA committed:
Fix a bug in j-job of atmos_prep

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_ATMOS_PREP

Refs: #94

XianwuXue-NOAA committed:
Make the vertical structure as default value

	The current g-w can not be used as flat structure for GEFS
		so I change the default structure is vertical structure

 On branch feature/v13_atmos_fcst
	modified:   rocoto/compile_install_all.sh

Refs: #94

XianwuXue-NOAA committed:
Improve the compile_install_all.sh

 On branch feature/v13_atmos_fcst
	modified:   rocoto/compile_install_all.sh

Refs: #94

XianwuXue-NOAA committed:
Reformat config.fcst

 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.fcst

Refs: #94

XianwuXue-NOAA committed:
Add VEGTYPE fix file back since p02 failed on hera
  after delete them

 On branch feature/v13_atmos_fcst
 Changes to be committed:
	modified:   scripts/exgefs_forecast.sh
Refs: #94

XianwuXue-NOAA committed:
Add SEND=NO back

 On branch feature/v13_atmos_fcst
 Changes to be committed:
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Move some vars from ex-script to config.fcst

 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.base
	modified:   parm/config/config.fcst
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Fix a minor bug

 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.fcst

Refs: #94

XianwuXue-NOAA committed:
Move vars from ex-script of fcst to config.fcst
 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.fcst
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @aerorahul I have moved some vars from ex-script of fcst to config.fcst and also did other modification. Please review my changes. Thanks

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @aerorahul When you have time, could you review my changes for this PR? Thanks.

XianwuXue-NOAA committed:
Update g-w commit

 On branch feature/v13_atmos_fcst
	modified:   sorc/checkout.sh

Refs: #94

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,86 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
-export GESIN=${GESIN:-$(compath.py $envir/com/$NET/${ver})/${RUN}.${PDY}/$cyc/nwges}
-export GESOUT=${GESOUT:-$(compath.py -o $NET/${ver})/${RUN}.${PDY}/$cyc/nwges}
+ver=${gefs_ver:0:5}
+export ROTDIR=${ROTDIR:-$(compath.py $envir/com/$NET/${ver})}
```
This should be defined before the j-job is called, and definitely should not use `compath.py`. `compath.py` should only be used for external packages (GFS, tracker, etc.).

XianwuXue-NOAA replied:
Is it the NCO new requirements? Because all the compath.py are in j-job in the current production.

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -26,8 +19,9 @@ cd $DATA
-export APRUN=${gefsmpexec:-"mpiexec -n $total_tasks"}
+#export total_tasks=${total_tasks:-320}
+#export APRUN=${gefsmpexec:-"mpiexec -n $total_tasks"}
+#export APRUN_FV3="mpiexec -n ${npe_fcst} -ppn ${npe_node_fcst} --cpu-bind depth --depth ${NTHREADS_FV3}"
```
Please remove the dead code.

XianwuXue-NOAA replied:
Will delete it in the next commit

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,86 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
 #############################
-configs="gefs gefs_wave gefs_fcst"
-config_path=$PARMgefs
+export CDATE=$PDY$cyc
```
We are avoiding using CDATE for the most part. I don't think this is even being used here after it is defined.
```suggestion
```

XianwuXue-NOAA replied:
The CDATE is used for the config.nsst in g-w. If there is no CDATE defined, then it will fail

WalterKolczynski-NOAA replied:
Update `config.nsst` to use `$PDY` and `$cyc`

XianwuXue-NOAA replied:
Will you update the config.nsst in g-w? I will not copy the config.nsst to GEFS repo.

WalterKolczynski-NOAA replied:
Just leave it then until we remove it in g-w

WalterKolczynski-NOAA commented on jobs/JGEFS_FORECAST:
```diff
@@ -71,51 +65,86 @@ export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
+  . ${config_path}/config.${config}
+  export err=$?
+  if [[ $err != 0 ]]; then
+    echo "FATAL ERROR in ${BASH_SOURCE}: Error while loading parm file ${config_path}/config.${config}"
```
`$BASH_SOURCE` is an array so should always specify an index, even if BASH defaults to the first element when none is provided (shellcheck will complain about this).
```suggestion
    echo "FATAL ERROR in ${BASH_SOURCE[0]}: Error while loading parm file ${config_path}/config.${config}"
```

XianwuXue-NOAA replied:
Will in the next commit

WalterKolczynski-NOAA commented on parm/config/config.base:
```diff
@@ -141,15 +140,23 @@ export PARMgefs=${PARMgefs:-${HOMEgefs}/parm}
 export EXPDIR=${PARMgfs}/config
-#export FIXgfs=${HOMEgfs}/fix
+export FIXgfs=${HOMEgfs}/fix
 #export USHgfs=${HOMEgfs}/ush
```
Remove all this dead code.

XianwuXue-NOAA replied:
OK. I will delete the dead code.

WalterKolczynski-NOAA commented on parm/config/config.base:
```diff
@@ -192,6 +216,78 @@ case "${CASE}" in
+
+    ;;
+  *)
+    echo "Unrecognized APP: ${1}"
```
```suggestion
    echo "Unrecognized APP: ${APP}"
```

XianwuXue-NOAA replied:
> Didn't complete the review of the config files. It seems like a lot of GEFS settings were tacked on to GFS settings without considering whether they are still needed or to update them to match the GFS style (for instance, using `$cplchm` for branching instead of `$DO_AERO`). Many settings have been updated for GFS v17/GEFS v13 and we should defer to the new settings as given in GFS unless there is a good reason to deviate (like turning on stochastic physics for some members). Please carefully review all of the settings and determine whether it is conflicting with an existing setting elsewhere. If you have any trouble determining whether a setting should be overridden for GEFS, I'll help and if necessary we'll ask MDAB.
> 
> Also, note the changes that were made in the [NOAA-EMC/global-workflow#1509](https://github.com/NOAA-EMC/global-workflow/pull/1509). You've updated the commit hash, but there are changes to the config files that will need to be duplicated here.

Thanks @WalterKolczynski-NOAA this PR is for atmos fcst job only, so I don't want to do the change related cplwav and cplchm. I will do the changes when I work on the fcst for wave and/or aerosol.
Regarding the config files, I will compare them in g-w and do the corresponding changes.

WalterKolczynski-NOAA commented on parm/config/config.base:
```diff
@@ -192,6 +216,78 @@ case "${CASE}" in
+export FHCYC=24
+
+# IAU related parameters
+export DOIAU=${DOIAU:-"NO"} # Enable 4DIAU for control with 3 increments
+if [[ $DOIAU == "NO" ]]; then
+  export IAU_OFFSET=0
+  export IAU_DELTHRS=0
+  export IAU_FHROT=0
+else
+  export IAUFHRS="3,6,9"
+  export IAU_FHROT=$(echo $IAUFHRS | cut -c1)
+  export IAU_DELTHRS=6
+  export IAU_OFFSET=6
+fi
```
GEFS has no DA of its own, does this need to be included?

XianwuXue-NOAA replied:
Maybe only need the DOIAU = NO, since I was told the GEFS will not use IAU. However, I am not sure whether it will be changed. So I keep these statements from config.base from g-w. In the future, we can delete unused statements.

XianwuXue-NOAA replied:
I will delete the DOIAU=YES condition.

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+#export SUFFIX=".nc"
+#export NEMSIO_IN=".false."
+#export NETCDF_IN=".true."
+#-----
```
Remove dead code

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+# if these variables are re-defined there.
+# Otherwise, the ensemble forecast will inherit from config.fcst
+
+echo "END: config.fcst"
```
But this is not the end of `config.fcst`

XianwuXue-NOAA replied:
This is just a reference of the Code segment from config.fcst in g-w, then we can easily compare the two config files. We can finalize it after the coupled fcst or aero fcst is ready.

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+export assim_freq=$gefs_cych
+
+# Settings for coupling with Wavewatch III
+export cplwav=${cplwav:-.true.}
```
This will override the `$APP` setting up-script.

XianwuXue-NOAA replied:
Yes, however, I don't want to change the cplwav and cplchm until I work on the fcst with wave and/or cplchm. Otherwise, I can not test them to see whether I did the right changes or not.

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+#
+# UPP parameters for GEFS
+#
+R2=$(echo $RUNMEM|cut -c4-5)
```
```suggestion
R2=${RUNMEM:3:2}
```

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+if [[ "$DO_AERO" = "YES" ]]; then # temporary settings for aerosol coupling
+    export AERO_DIAG_TABLE="${AERO_DIAG_TABLE:-$HOMEgfs/parm/parm_fv3diag/diag_table.aero}"
+    export AERO_FIELD_TABLE="${AERO_FIELD_TABLE:-$HOMEgfs/parm/parm_fv3diag/field_table.aero}"
+    export AERO_EMIS_FIRE=$( echo "${AERO_EMIS_FIRE:-none}" | awk '{ print tolower($1) }' )
```
```suggestion
    export AERO_EMIS_FIRE="${AERO_EMIS_FIRE,,}"
```

XianwuXue-NOAA replied:
This is the copy from config.fcst in g-w. I don't want to change the scripts related DO_AERO until I do the aerol-fcst job. If you want to change it, please change it in config.fcst. I will do the update.

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+fi
+
+# Choose coupling with wave
+if [[ "$DO_WAVE" = "YES" && "$WAVE_CDUMP" != "gdas" ]]; then
```
```suggestion
if [[ "$DO_WAVE" = "YES" ]]; then
```

XianwuXue-NOAA replied:
This is the copy of code segments from config.fcst in g-w, I will follow the the config.fcst of g-w changes

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+export WAV_MOD_ID=${WAV_MOD_ID:-gefs}
+export coupling_interval_sec=${coupling_interval_sec:-3600}
+
+if [[ $cplchm = ".true." ]]; then
+  export DELTIM=300 # Needed to match setting used by GSD
+
+  if [[ $AEROSOL_INIT_TYPE == "warm" ]]; then
+    export warm_start=".true."
+    export read_increment=".true."
+  fi
+
+  export DIAG_TABLE=${DIAG_TABLE:-$HOMEgfs/parm/parm_fv3diag/chm_diag_table}
+
+  export lgfdlmprad=".true."
+  export effr_in=".true."
+  export USE_COUPLER_RES="YES"
+fi
```
I think this has all been superseded by other code/settings coming over from GFS.
```suggestion
```

XianwuXue-NOAA replied:
Since I am not working on aerosol-fcst job, so I don't want to include the changes for chem in atmos fcst

WalterKolczynski-NOAA commented on parm/config/config.fcst:
```diff
@@ -0,0 +1,654 @@
+
+export FHCYC=${FHCYC:-24}
+
+#  The VEGTYPE fix file:
+export FIX_AM=${FIX_AM:-${FIXgfs}/am}
+export FNVETC=${FNVETC:-${FIX_AM}/global_vegtype.igbp.t$MTNRSL.rg.grb}
+export FNTSFC=${FNTSFC:-${FIX_AM}/RTGSST.1982.2012.monthly.clim.grb}
+export FNAISC=${FNAISC:-${FIX_AM}/CFSR.SEAICE.1982.2012.monthly.clim.grb}
+export FNABSC=${FNABSC:-${FIX_AM}/global_mxsnoalb.uariz.t$MTNRSL.rg.grb}
+export FNALBC=$FIX_AM/global_snowfree_albedo.bosu.t$MTNRSL.rg.grb
+export FNALBC2=$FIX_AM/global_albedo4.1x1.grb
+export FNSMCC=$FIX_AM/global_soilmgldas.t$MTNRSL.grb
+export FNSOTC=$FIX_AM/global_soiltype.statsgo.t$MTNRSL.rg.grb
```
These are all defined in forecast_postdet and we should use the values there, as they have been updated for the newer model.

XianwuXue-NOAA replied:
after I removed these variables here, then the fcst will fail. So I add them back. I don't know who will test it, could you please let others to know this issue. Because someone resolve this issue, I think it is better to keep them, otherwise, the fcst will fail. @WalterKolczynski-NOAA @aerorahul

WalterKolczynski-NOAA replied:
IIRC, you had one member fail because of instability. That doesn't mean we should be using the old fix files.

XianwuXue-NOAA replied:
I have removed them and ran it on WCOSS2 successfully (c00, p01, p02)

WalterKolczynski-NOAA commented on rocoto/bin/hera/forecast_hr.sh:
```diff
@@ -1,24 +1,54 @@
+ver=${gefs_ver%${gefs_ver##v+([0-9]).+([0-9])}}
+shopt -u extglob
 
+export ROTDIR=${COMROOT}/gefs/${ver}
```
For development, I don't think we need to add the additional unnecessary directory levels that NCO is.

```suggestion
# Assuming $EXPDIR is already built into $COMROOT
ROTDIR=${COMROOT}
```

XianwuXue-NOAA replied:
GEFS will not use EXPDIR, This line is to set ROTDIR because I call j-job. This is only be used on Hera.

WalterKolczynski-NOAA commented on rocoto/bin/hera/forecast_hr.sh:
```diff
@@ -1,24 +1,54 @@
+shopt -u extglob
 
+export ROTDIR=${COMROOT}/gefs/${ver}
+export ROTDIR_GFS=${HOMEdata}/gfs/${gfs_ver}
```
Same here. For GFS development we don't include the extra directory levels NCO does. For operations, this needs to use `compath.py`, since GFS is an external package.

XianwuXue-NOAA replied:
Yes, however, this is only used on Hera. On wcos2, I will use compath.py to get the path, but it can not used on Hera.

WalterKolczynski-NOAA reviewed (changes requested):
Didn't complete the review of the config files. It seems like a lot of GEFS settings were tacked on to GFS settings without considering whether they are still needed or to update them to match the GFS style (for instance, using `$cplchm` for branching instead of `$DO_AERO`). Many settings have been updated for GFS v17/GEFS v13 and we should defer to the new settings as given in GFS unless there is a good reason to deviate (like turning on stochastic physics for some members). Please carefully review all of the settings and determine whether it is conflicting with an existing setting elsewhere. If you have any trouble determining whether a setting should be overridden for GEFS, I'll help and if necessary we'll ask MDAB.

Also, note the changes that were made in the NOAA-EMC/global-workflow#1509. You've updated the commit hash, but there are changes to the config files that will need to be duplicated here.

XianwuXue-NOAA committed:
Modify based on Walter's comments and g-w changes

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   parm/config/config.base
	modified:   parm/config/config.fcst
	modified:   parm/config/config.resources
	modified:   parm/config/config.ufs

Refs: #94

XianwuXue-NOAA committed:
Modify based on reviewer's comments

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	modified:   parm/config/config.base

Refs: #94

XianwuXue-NOAA committed:
Modify the config files

 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.fcst
	modified:   parm/config/config.ufs
	modified:   scripts/exgefs_forecast.sh

Refs: #94

XianwuXue-NOAA committed:
Fix a bug to run fcst

 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.resources

Refs: #94

XianwuXue-NOAA committed:
Remove the dead codes

 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.fcst

Refs: #94

XianwuXue-NOAA committed:
Seperate config.fcst to config.fcst & config.efcs

 On branch feature/v13_atmos_fcst
	modified:   jobs/JGEFS_FORECAST
	new file:   parm/config/config.efcs
	modified:   parm/config/config.fcst

Refs: #94

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @aerorahul I have decided the config.fcst into two files: config.fcst and config.efcs. Which will shorten the config.fcst. Please re-review my changes for this PR. I think I have resolved all the comments. If I missed something, please let me know.

XianwuXue-NOAA committed:
Fix a bug

 On branch feature/v13_atmos_fcst
	modified:   parm/config/config.efcs

Refs: #94

XianwuXue-NOAA commented:
Since the GEFSv13 workflow will be in g-w, I will merge this PR to close this PR and for the future's reference. I will do the same PR for the other jobs. If you have any suggestions, please let me know. @WalterKolczynski-NOAA @aerorahul
